File: backend/src/ports/input/api/v1/public.py
"""공개 API — 인증 없음.

블로그 플랫폼 트래픽의 대부분은 로그인하지 않은 독자와 검색 크롤러다.
이 라우터는 절대 인증을 요구하지 않으며, 반환값에 비공개 정보를 담지 않는다.
"""

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from pydantic import BaseModel

from src.ports.input.api.v1.dependencies import (
    get_block_repo,
    get_comment_repo,
    get_follow_repo,
    get_like_repo,
    get_optional_user_id,
    get_post_repo,
    get_series_repo,
    get_tag_repo,
    get_user_repo,
    viewer_key,
)
from src.ports.output.repositories.moderation_repository import BlockRepository
from src.ports.output.repositories.post_repository import PostRepository
from src.ports.output.repositories.social_repository import (
    CommentRepository,
    FollowRepository,
    LikeRepository,
)
from src.ports.output.repositories.taxonomy_repository import SeriesRepository, TagRepository
from src.ports.output.repositories.user_repository import UserRepository

router = APIRouter()

MAX_PAGE_SIZE = 50


def _author(user) -> Dict[str, Any]:
    """공개해도 되는 작성자 정보만. 이메일은 절대 나가지 않는다."""
    return {
        "handle": user.handle,
        "display_name": user.display_name or user.handle,
        "avatar_url": user.avatar_url,
        "bio": user.bio,
    }


def _post_card(post) -> Dict[str, Any]:
    """목록용. 본문 전체를 싣지 않는다."""
    return {
        "id": post.id,
        "slug": post.slug,
        "title": post.title,
        "summary": post.summary,
        "cover_url": post.cover_url,
        "published_at": post.published_at.isoformat() if post.published_at else None,
        "like_count": post.like_count,
        "comment_count": post.comment_count,
        "tags": [link.tag.display_name for link in post.tags],
        "author": _author(post.user),
        "url": f"/@{post.user.handle}/{post.slug}",
    }


class PostList(BaseModel):
    items: List[Dict[str, Any]]
    has_more: bool


def _paged(items: List, limit: int) -> Dict[str, Any]:
    # limit+1 을 조회해 다음 페이지 존재 여부를 판단한다 (전체 COUNT 를 피한다).
    return {"items": [_post_card(p) for p in items[:limit]], "has_more": len(items) > limit}


# ------------------------------------------------------------------- 피드


# 트렌딩 기간 선택지 (Velog 의 "이번 주" 드롭다운에 해당)
PERIOD_DAYS = {"week": 7, "month": 30, "year": 365, "all": None}


@router.get("/feed", response_model=PostList)
def feed(
    sort: str = Query("recent", pattern="^(recent|trending|recommended|following)$"),
    period: str = Query("week", pattern="^(week|month|year|all)$"),
    tag: Optional[str] = None,
    limit: int = Query(20, ge=1, le=MAX_PAGE_SIZE),
    offset: int = Query(0, ge=0),
    viewer_id: Optional[str] = Depends(get_optional_user_id),
    post_repo: PostRepository = Depends(get_post_repo),
    block_repo: BlockRepository = Depends(get_block_repo),
):
    """홈 피드.

    - recent      : 최신순
    - trending    : 반응 가중치 + 기간 필터
    - recommended : 내가 좋아요한 글의 태그와 겹치는 글 (로그인 필요, 없으면 트렌딩)
    - following   : 내가 팔로우한 사람의 글 (로그인 필요, 없으면 최신)
    """
    # 차단한 사람의 글은 내 화면에서 빠져야 한다.
    blocked = block_repo.blocked_ids(viewer_id) if viewer_id else []

    if sort == "following":
        if not viewer_id:
            return _paged(post_repo.list_feed(limit=limit + 1, offset=offset), limit)
        return _paged(post_repo.list_following_feed(viewer_id, limit + 1, offset), limit)

    if sort == "recommended":
        if viewer_id:
            posts = post_repo.list_recommended(viewer_id, limit + 1, offset)
            if posts:
                return _paged(posts, limit)
        # 신호가 없으면 트렌딩으로 대체한다. 빈 화면을 보여주는 것보다 낫다.
        sort = "trending"

    posts = post_repo.list_feed(
        limit=limit + 1,
        offset=offset,
        sort=sort,
        tag=tag,
        # 인기순은 최근 글 중에서만 뽑는다. 아니면 오래된 글이 상단을 영구 점유한다.
        since_days=PERIOD_DAYS[period] if sort == "trending" else None,
        exclude_user_ids=blocked,
    )
    return _paged(posts, limit)


@router.get("/search", response_model=PostList)
def search(
    q: str = Query(min_length=1, max_length=100),
    limit: int = Query(20, ge=1, le=MAX_PAGE_SIZE),
    offset: int = Query(0, ge=0),
    post_repo: PostRepository = Depends(get_post_repo),
):
    return _paged(post_repo.search(q, limit + 1, offset), limit)


@router.get("/tags")
def popular_tags(
    limit: int = Query(30, ge=1, le=100),
    tag_repo: TagRepository = Depends(get_tag_repo),
):
    return [
        {"name": t.name, "display_name": t.display_name, "post_count": t.post_count}
        for t in tag_repo.list_popular(limit)
    ]


# --------------------------------------------------------------- 블로그 / 글


@router.get("/blogs/{handle}")
def blog_home(
    handle: str,
    viewer_id: Optional[str] = Depends(get_optional_user_id),
    user_repo: UserRepository = Depends(get_user_repo),
    post_repo: PostRepository = Depends(get_post_repo),
    series_repo: SeriesRepository = Depends(get_series_repo),
    follow_repo: FollowRepository = Depends(get_follow_repo),
):
    user = user_repo.get_by_handle(handle)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="블로그를 찾을 수 없습니다.")

    return {
        **_author(user),
        "post_count": post_repo.count_by_user(user.id, only_published=True),
        "follower_count": user.follower_count,
        "following_count": user.following_count,
        "series": [
            {"slug": s.slug, "name": s.name, "post_count": len(s.posts)}
            for s in series_repo.list_by_user(user.id)
        ],
        # 로그인한 사람이 볼 때만 채워지는 개인화 값
        "is_following": bool(viewer_id) and follow_repo.exists(viewer_id, user.id),
        "is_me": viewer_id == user.id,
    }


@router.get("/blogs/{handle}/posts", response_model=PostList)
def blog_posts(
    handle: str,
    limit: int = Query(20, ge=1, le=MAX_PAGE_SIZE),
    offset: int = Query(0, ge=0),
    user_repo: UserRepository = Depends(get_user_repo),
    post_repo: PostRepository = Depends(get_post_repo),
):
    user = user_repo.get_by_handle(handle)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="블로그를 찾을 수 없습니다.")
    posts = post_repo.list_by_user(user.id, only_published=True, limit=limit + 1, offset=offset)
    return _paged(posts, limit)


@router.get("/blogs/{handle}/posts/{slug}")
def post_detail(
    request: Request,
    handle: str,
    slug: str,
    viewer_id: Optional[str] = Depends(get_optional_user_id),
    post_repo: PostRepository = Depends(get_post_repo),
    like_repo: LikeRepository = Depends(get_like_repo),
    follow_repo: FollowRepository = Depends(get_follow_repo),
    comment_repo: CommentRepository = Depends(get_comment_repo),
    series_repo: SeriesRepository = Depends(get_series_repo),
):
    post = post_repo.get_public(handle, slug)
    if not post:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="글을 찾을 수 없습니다.")

    # 같은 사람의 재방문은 24시간 안에서 한 번만 센다. 새로고침으로 오르는 숫자는
    # 아무 의미가 없고, 그 숫자를 근거로 하는 트렌딩 정렬까지 같이 망가진다.
    post_repo.record_view(post.id, viewer_key(request, viewer_id), viewer_id)

    series_context = series_repo.context_for_post(post.id)

    comments = comment_repo.list_for_post(post.id)
    roots = [c for c in comments if c.parent_id is None]
    replies: Dict[str, List] = {}
    for c in comments:
        if c.parent_id:
            replies.setdefault(c.parent_id, []).append(c)

    def render(comment) -> Dict[str, Any]:
        deleted = comment.deleted_at is not None
        return {
            "id": comment.id,
            # 답글이 달린 댓글은 자리를 남긴다. 내용만 가린다.
            "body": None if deleted else comment.body,
            "deleted": deleted,
            "created_at": comment.created_at.isoformat() if comment.created_at else None,
            "author": None if deleted else _author(comment.user),
            "is_mine": bool(viewer_id) and comment.user_id == viewer_id,
            "replies": [render(r) for r in replies.get(comment.id, [])],
        }

    return {
        **_post_card(post),
        "content_md": post.content_md,
        "view_count": post.view_count,
        "comments": [render(c) for c in roots],
        "is_liked": bool(viewer_id) and like_repo.exists(post.id, viewer_id),
        "is_following_author": bool(viewer_id) and follow_repo.exists(viewer_id, post.user_id),
        "is_mine": viewer_id == post.user_id,
        "series": _series_nav(series_context, handle),
    }


def _series_nav(context: Optional[Dict[str, Any]], handle: str) -> Optional[Dict[str, Any]]:
    """시리즈 앞뒤 네비게이션. 시리즈에 속하지 않으면 None."""
    if not context:
        return None

    def link(p) -> Optional[Dict[str, Any]]:
        return {"title": p.title, "url": f"/@{handle}/{p.slug}"} if p else None

    series = context["series"]
    return {
        "name": series.name,
        "url": f"/@{handle}/series/{series.slug}",
        "position": context["position"],
        "total": context["total"],
        "previous": link(context["previous"]),
        "next": link(context["next"]),
    }


@router.get("/blogs/{handle}/series/{series_slug}")
def series_detail(
    handle: str,
    series_slug: str,
    user_repo: UserRepository = Depends(get_user_repo),
    series_repo: SeriesRepository = Depends(get_series_repo),
):
    series = series_repo.get_public(handle, series_slug)
    if not series:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="시리즈를 찾을 수 없습니다.")

    user = user_repo.get_by_handle(handle)
    ordered = [link.post for link in series.posts if link.post is not None]
    return {
        "slug": series.slug,
        "name": series.name,
        "description": series.description,
        "author": _author(user),
        "items": [_post_card(p) for p in ordered],
    }


# ------------------------------------------------------------------- SEO


@router.get("/sitemap")
def sitemap(post_repo: PostRepository = Depends(get_post_repo)):
    """프론트의 sitemap.xml 생성에 쓸 원본 목록."""
    return [
        {
            "url": f"/@{p.user.handle}/{p.slug}",
            "updated_at": (p.updated_at or p.published_at).isoformat()
            if (p.updated_at or p.published_at)
            else None,
        }
        for p in post_repo.all_published_for_sitemap()
        if p.user and p.user.handle
    ]


@router.get("/blogs/{handle}/rss")
def blog_rss_source(
    handle: str,
    user_repo: UserRepository = Depends(get_user_repo),
    post_repo: PostRepository = Depends(get_post_repo),
):
    """RSS 생성용 원본. 우리가 남의 RSS 를 읽던 입장에서 이제 내보내는 쪽이 된다."""
    user = user_repo.get_by_handle(handle)
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="블로그를 찾을 수 없습니다.")

    posts = post_repo.list_by_user(user.id, only_published=True, limit=30, offset=0)
    return {
        "author": _author(user),
        "items": [
            {
                "title": p.title,
                "url": f"/@{user.handle}/{p.slug}",
                "summary": p.summary,
                "published_at": p.published_at.isoformat() if p.published_at else None,
            }
            for p in posts
        ],
    }

File: frontend/lib/api/public.ts
/**
 * 공개 콘텐츠 조회 — 서버 전용.
 *
 * 이 경로만큼은 클라이언트에서 베어러 토큰으로 가져오면 안 된다.
 * 검색 크롤러와 SNS 미리보기 봇은 로그인도 없고 JS 실행도 기대할 수 없기 때문에,
 * 공개 글은 반드시 서버에서 렌더링돼야 색인된다.
 *
 * 캐시는 태그 기반으로 무효화한다 (발행 시 revalidateTag).
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export const SITE_URL = process.env.NEXT_PUBLIC_SITE_URL || 'http://localhost:3000'
export const SITE_NAME = 'Devshiplog'

export interface Author {
  handle: string
  display_name: string
  avatar_url: string | null
  bio: string | null
}

export interface PostStack {
  name: string
  version: string | null
}

export interface Freshness {
  /** fresh: 최근 검증 · aging: 확인 필요 · stale: 낡음 · unverified: 검증 이력 없음 */
  level: 'fresh' | 'aging' | 'stale' | 'unverified'
  days_since_verified: number | null
  outdated: { name: string; version: string; latest_major: number }[]
  /** 화면에 그대로 쓸 수 있는 한 줄 설명 */
  reason: string
  verified_at: string | null
}

export interface PostCard {
  id: string
  slug: string
  title: string
  summary: string | null
  cover_url: string | null
  stacks: PostStack[]
  freshness: Freshness
  published_at: string | null
  like_count: number
  comment_count: number
  tags: string[]
  author: Author
  url: string
}

export interface CommentNode {
  id: string
  body: string | null
  deleted: boolean
  created_at: string | null
  author: Author | null
  is_mine: boolean
  replies: CommentNode[]
}

export interface SeriesNavData {
  name: string
  url: string
  /** 공개된 글만 세어서 매긴 순번. 중간 편을 내려도 번호가 건너뛰지 않는다. */
  position: number
  total: number
  previous: { title: string; url: string } | null
  next: { title: string; url: string } | null
}

export interface PostDetail extends PostCard {
  content_md: string
  view_count: number
  comments: CommentNode[]
  is_liked: boolean
  is_following_author: boolean
  is_mine: boolean
  series: SeriesNavData | null
  signals: { works: number; broken: number; my_signal: 'works' | 'broken' | null }
}

export interface BlogHome extends Author {
  post_count: number
  follower_count: number
  following_count: number
  series: { slug: string; name: string; post_count: number }[]
  is_following: boolean
  is_me: boolean
}

export interface PostList {
  items: PostCard[]
  has_more: boolean
}

/** 캐시 태그 — 발행/수정 시 이 태그를 무효화한다. */
export const cacheTags = {
  feed: 'feed',
  blog: (handle: string) => `blog:${handle}`,
  post: (handle: string, slug: string) => `post:${handle}:${slug}`,
}

interface FetchOptions {
  tags?: string[]
  revalidate?: number
}

async function getJson<T>(path: string, options: FetchOptions = {}): Promise<T | null> {
  try {
    const res = await fetch(`${API_BASE_URL}/api/v1/public${path}`, {
      next: {
        // 발행 시 태그로 즉시 무효화하되, 안전망으로 시간 기반 재검증도 둔다.
        revalidate: options.revalidate ?? 60,
        tags: options.tags,
      },
    })
    if (res.status === 404) return null
    if (!res.ok) {
      console.error(`public API ${path} → ${res.status}`)
      return null
    }
    return (await res.json()) as T
  } catch (error) {
    // 백엔드가 죽어도 페이지는 "없음" 으로 뜨는 편이 500 보다 낫다.
    console.error(`public API ${path} 실패:`, error)
    return null
  }
}

export type FeedSort = 'recent' | 'trending' | 'recommended' | 'following'
export type FeedPeriod = 'week' | 'month' | 'year' | 'all'

export function getFeed(
  params: {
    sort?: FeedSort
    period?: FeedPeriod
    tag?: string
    limit?: number
    offset?: number
  } = {}
) {
  const q = new URLSearchParams()
  if (params.sort) q.set('sort', params.sort)
  if (params.period) q.set('period', params.period)
  if (params.tag) q.set('tag', params.tag)
  q.set('limit', String(params.limit ?? 20))
  q.set('offset', String(params.offset ?? 0))
  return getJson<PostList>(`/feed?${q}`, { tags: [cacheTags.feed], revalidate: 30 })
}

export interface StackSummary {
  name: string
  post_count: number
}

export function getPopularStacks(limit = 40) {
  return getJson<StackSummary[]>(`/stacks?limit=${limit}`, {
    tags: [cacheTags.feed],
    revalidate: 300,
  })
}

export type StackSort = 'fresh_first' | 'recent' | 'trending'

/**
 * 이 스택으로 쓰인 글.
 *
 * 기본 정렬이 fresh_first 인 것이 이 제품의 입장이다.
 * 최신순으로 두면 "최근에 쓴 낡은 글" 이 위로 온다. 독자가 원하는 건
 * 최근에 쓴 글이 아니라 지금도 되는 글이다.
 */
export function getPostsByStack(
  name: string,
  params: { version?: string; sort?: StackSort; limit?: number } = {}
) {
  const q = new URLSearchParams()
  if (params.version) q.set('version', params.version)
  q.set('sort', params.sort ?? 'fresh_first')
  q.set('limit', String(params.limit ?? 24))
  return getJson<PostList>(`/stacks/${encodeURIComponent(name)}?${q}`, {
    tags: [cacheTags.feed],
    revalidate: 60,
  })
}

export function searchPosts(query: string, limit = 20) {
  return getJson<PostList>(`/search?q=${encodeURIComponent(query)}&limit=${limit}`, {
    revalidate: 0,
  })
}

export function getPopularTags(limit = 30) {
  return getJson<{ name: string; display_name: string; post_count: number }[]>(
    `/tags?limit=${limit}`,
    { tags: [cacheTags.feed], revalidate: 300 }
  )
}

export function getBlog(handle: string) {
  return getJson<BlogHome>(`/blogs/${encodeURIComponent(handle)}`, {
    tags: [cacheTags.blog(handle)],
  })
}

export function getBlogPosts(handle: string, limit = 20, offset = 0) {
  return getJson<PostList>(
    `/blogs/${encodeURIComponent(handle)}/posts?limit=${limit}&offset=${offset}`,
    { tags: [cacheTags.blog(handle)] }
  )
}

export function getPost(handle: string, slug: string) {
  return getJson<PostDetail>(
    `/blogs/${encodeURIComponent(handle)}/posts/${encodeURIComponent(slug)}`,
    { tags: [cacheTags.post(handle, slug)] }
  )
}

export function getSeries(handle: string, seriesSlug: string) {
  return getJson<{
    slug: string
    name: string
    description: string | null
    author: Author
    items: PostCard[]
  }>(`/blogs/${encodeURIComponent(handle)}/series/${encodeURIComponent(seriesSlug)}`, {
    tags: [cacheTags.blog(handle)],
  })
}

export function getSitemapEntries() {
  // 발행하면 사이트맵도 함께 갱신돼야 새 글이 바로 색인 대상이 된다.
  return getJson<{ url: string; updated_at: string | null }[]>('/sitemap', {
    tags: [cacheTags.feed],
    revalidate: 3600,
  })
}

export function getRssSource(handle: string) {
  return getJson<{
    author: Author
    items: { title: string; url: string; summary: string | null; published_at: string | null }[]
  }>(`/blogs/${encodeURIComponent(handle)}/rss`, { tags: [cacheTags.blog(handle)] })
}

/**
 * 업로드 주소를 절대 URL 로.
 *
 * 백엔드는 '/uploads/...' 상대 경로를 준다(도메인이 바뀌어도 DB 가 안 깨지게).
 * 하지만 OG 태그·JSON-LD 처럼 남의 서버가 읽는 자리에는 절대 주소가 필요하다.
 */
export function absoluteUrl(path: string): string {
  return /^https?:\/\//.test(path) ? path : `${SITE_URL}${path.startsWith('/') ? '' : '/'}${path}`
}

/** 날짜를 화면용 문자열로. 서버·클라이언트 렌더 결과가 어긋나지 않도록 UTC 고정. */
export function formatDate(iso: string | null): string {
  if (!iso) return ''
  const d = new Date(iso)
  return `${d.getUTCFullYear()}. ${d.getUTCMonth() + 1}. ${d.getUTCDate()}.`
}
